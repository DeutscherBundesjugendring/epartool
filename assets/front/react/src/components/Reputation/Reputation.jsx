import React from 'react';
import ThumbButton from '../../components/ThumbButton/ThumbButton';

/* global baseUrl */
/* global followupTranslations */

class Reputation extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      voted: props.voted,
      expectLikesUpdate: false,
      votingLimitErrorShowed: false,
      initLikeCount: props.likeCount,
      initDislikeCount: props.dislikeCount,
    };

    this.like = this.like.bind(this);
    this.dislike = this.dislike.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    // jiri@visionapps.cz 30.05.2017 DBJR-1167
    // Preventing of repeated voting is implemented on backend.
    // This construction is only for informing user and it is not fully reliable in case of
    // parallel voting from more users at the same time
    if (this.state.expectLikesUpdate && !this.state.votingLimitErrorShowed
      && nextProps.likeCount === this.state.initLikeCount
      && nextProps.dislikeCount === this.state.initDislikeCount) {
      this.setState({ votingLimitErrorShowed: true });
      alert(followupTranslations.votingLimitError);
    }
  }

  like() {
    this.props.likeAction();
    this.setState({ voted: true, expectLikesUpdate: true });
  }

  dislike() {
    this.props.dislikeAction();
    this.setState({ voted: true, expectLikesUpdate: true });
  }

  render() {
    return (
      <div>
        <div style={{ display: 'inline-block' }}>
          <span className="badge offset-right-small">{this.props.likeCount}</span>
          <ThumbButton type="like" onTouchTap={this.like} disabled={this.state.voted} />
          <div className="text-center offset-top-small small">Like</div>
        </div>
        <div style={{ display: 'inline-block' }}>
          <span className="badge offset-left offset-right-small">{this.props.dislikeCount}</span>
          <ThumbButton type="dislike" onTouchTap={this.dislike} disabled={this.state.voted} />
          <div className="text-center offset-top-small small">Dislike</div>
        </div>
      </div>
    );
  }
}

Reputation.defaultProps = {
  voted: false,
};

Reputation.propTypes = {
  likeCount: React.PropTypes.number.isRequired,
  dislikeCount: React.PropTypes.number.isRequired,
  likeAction: React.PropTypes.func.isRequired,
  dislikeAction: React.PropTypes.func.isRequired,
  voted: React.PropTypes.bool,
};

export default Reputation;
