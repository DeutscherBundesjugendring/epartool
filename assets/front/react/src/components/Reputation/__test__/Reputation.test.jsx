import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';
import React from 'react';
import injectTapEventPlugin from 'react-tap-event-plugin';
import Reputation from '../Reputation';


injectTapEventPlugin();

const element = (
    <Reputation
        likeAction={() => {}}
        likeCount={0}
        dislikeAction={() => {}}
        dislikeCount={0}
    />
);

describe('rendering', () => {
  it('renders correctly', () => {
    const tree = shallow(
      React.cloneElement(element)
    );

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly disabled voting', () => {
    const tree = shallow(
        React.cloneElement(element)
    );

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});

describe('functionality', () => {
    it('calls like action', () => {
        const spy = sinon.spy();
        const component = shallow(
            React.cloneElement(element, { likeAction: spy })
        );

        component.find('ThumbButton').first().simulate('touchTap', { preventDefault: () => {} });
        expect(spy.calledOnce).toEqual(true);
    });

    it('calls dislike action', () => {
        const spy = sinon.spy();
        const component = shallow(
            React.cloneElement(element, { dislikeAction: spy })
        );

        component.find('ThumbButton').at(1).simulate('touchTap', { preventDefault: () => {} });
        expect(spy.calledOnce).toEqual(true);
    });
});
