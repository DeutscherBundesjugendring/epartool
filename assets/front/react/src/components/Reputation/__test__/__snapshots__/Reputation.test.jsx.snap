// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly 1`] = `
<div>
  <div
    style={
      Object {
        "display": "inline-block",
      }
    }
  >
    <span
      className="badge offset-right-small"
    >
      0
    </span>
    <ThumbButton
      disabled={false}
      onTouchTap={[Function]}
      type="like"
    />
    <div
      className="text-center offset-top-small small"
    >
      Like
    </div>
  </div>
  <div
    style={
      Object {
        "display": "inline-block",
      }
    }
  >
    <span
      className="badge offset-left offset-right-small"
    >
      0
    </span>
    <ThumbButton
      disabled={false}
      onTouchTap={[Function]}
      type="dislike"
    />
    <div
      className="text-center offset-top-small small"
    >
      Dislike
    </div>
  </div>
</div>
`;

exports[`rendering renders correctly disabled voting 1`] = `
<div>
  <div
    style={
      Object {
        "display": "inline-block",
      }
    }
  >
    <span
      className="badge offset-right-small"
    >
      0
    </span>
    <ThumbButton
      disabled={false}
      onTouchTap={[Function]}
      type="like"
    />
    <div
      className="text-center offset-top-small small"
    >
      Like
    </div>
  </div>
  <div
    style={
      Object {
        "display": "inline-block",
      }
    }
  >
    <span
      className="badge offset-left offset-right-small"
    >
      0
    </span>
    <ThumbButton
      disabled={false}
      onTouchTap={[Function]}
      type="dislike"
    />
    <div
      className="text-center offset-top-small small"
    >
      Dislike
    </div>
  </div>
</div>
`;
